specs/IFramePermissionRequested.md

```diff
@@ -0,0 +1,363 @@
+                                    return S_OK;
+                            }).Get(),
+                            &m_PermissionRequestedToken));
+                    }
+                    return S_OK;
+            }).Get(),
+            &m_FrameCreatedToken));
+    }
+}
+
+static PCWSTR NameOfPermissionKind(COREWEBVIEW2_PERMISSION_KIND kind)
+{
+    switch (kind)
+    {
+        case COREWEBVIEW2_PERMISSION_KIND_MICROPHONE:
+            return L"Microphone";
+        case COREWEBVIEW2_PERMISSION_KIND_CAMERA:
+            return L"Camera";
+        case COREWEBVIEW2_PERMISSION_KIND_GEOLOCATION:
+            return L"Geolocation";
+        case COREWEBVIEW2_PERMISSION_KIND_NOTIFICATIONS:
+            return L"Notifications";
+        case COREWEBVIEW2_PERMISSION_KIND_OTHER_SENSORS:
+            return L"Generic Sensors";
+        case COREWEBVIEW2_PERMISSION_KIND_CLIPBOARD_READ:
+            return L"Clipboard Read";
+        default:
+            return L"Unknown resources";
+    }
+}
+
+static void PutHandled(ICoreWebView2PermissionRequestedEventArgs* args)
+{
+    // In the case of an iframe requesting permission, the default behavior is
+    // to first raise the PermissionRequested event off of the CoreWebView2Frame
+    // and invoke it's handlers, and then raise the event off the CoreWebView2
+    // and invoke it's handlers. However, If we set Handled to true on the
+    // CoreWebView2Frame event handler, then we will not raise the
+    // PermissionRequested event off the CoreWebView2.
+    wil::com_ptr<ICoreWebView2PermissionRequestedEventArgs2> args2;
+    CHECK_FAILURE(args->QueryInterface(IID_PPV_ARGS(&args2)));
+    if (args2)
+    {
+        CHECK_FAILURE(args2->put_Handled(true));
+    }
+}
+```
+
+## C#: Registering IFrame Permission Requested Handler
+```c#
+private WebView2 m_webview;
+Dictionary<Tuple<string, CoreWebView2PermissionKind, bool>, bool> m_cachedPermissions;
+
+// Example Use Case - an app may want to specify that all iframes should
+// create a message box to prompt the user for approval on a permission request.
+// The approval state could be cached so that future requests are automatically
+// handled if they have been requested previously.
+void RegisterIFramePermissionRequestedHandler()
+{
+    m_webview.CoreWebView2.FrameCreated += (sender, frameCreatedArgs) =>
+    {
+        frameCreatedArgs.Frame.PermissionRequested += (frameSender, permissionArgs) =>
+        {
+            var cachedKey = Tuple.Create(permissionArgs.Uri,
+                permissionArgs.PermissionKind, permissionArgs.IsUserInitiated);
+
+            if (m_cachedPermissions.ContainsKey(cachedKey))
+            {
+                permissionArgs.State = m_cachedPermissions[cachedKey]
+                                            ? CoreWebView2PermissionState.Allow
+                                            : CoreWebView2PermissionState.Deny;
+
+                permissionArgs.Handled = true;
```

Comment:
.Handled will throw if the underlying WebView2 Runtime doesn't implement your new interface. However we only run this code if the CoreWebView2Frame.PermissionRequested event exists so its not a problem because they both exist together. But would be nice to add a comment about that.

Reply:
Created a PutHandled to match C++ and added the try catch. Added comment about how it shouldn't be a problem as well.